src/SourceCode.Clay.Json/Pointers/JsonPointer.cs

```diff
@@ -0,0 +1,289 @@
+                            else throw new InvalidOperationException($"The new index token ('{token.Value}') is not supported.");
+                            continue;
+                        }
+
+                        var a = (JsonArray)target;
+                        var index = token.ArrayIndex;
+                        if (!index.HasValue)
+                        {
+                            if (options.HasFlag(JsonPointerEvaluationOptions.ArrayMembersAreNull)) target = null;
+                            else throw new InvalidOperationException($"An array cannot be indexed by the member '{token.Value}'.");
+                            continue;
+                        }
+
+                        if (index.Value >= a.Count)
+                        {
+                            if (options.HasFlag(JsonPointerEvaluationOptions.InvalidIndiciesAreNull)) target = null;
+                            else throw new InvalidOperationException($"The index {index.Value} falls outside the array length ({a.Count}).");
+                            continue;
+                        }
+
+                        target = a[index.Value];
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Unsupported JSON type {target.GetType()}.");
+                }
+            }
+
+            return target;
+        }
+
+        #endregion
+
+        #region Equatable
+
+        /// <summary>Indicates whether this instance and a specified object are equal.</summary>
+        /// <param name="obj">The object to compare with the current instance.</param>
+        /// <returns>true if <paramref name="obj">obj</paramref> and this instance are the same type and represent the same value; otherwise, false.</returns>
+        public override bool Equals(object obj) => obj is JsonPointer o && Equals(o);
+
+        /// <summary>Indicates whether the current object is equal to another object of the same type.</summary>
+        /// <param name="other">An object to compare with this object.</param>
+        /// <returns>true if the current object is equal to the <paramref name="other">other</paramref> parameter; otherwise, false.</returns>
+        public bool Equals(JsonPointer other)
+        {
+            if (ReferenceEquals(_tokens, other._tokens)) return true;
+            if (ReferenceEquals(_tokens, null) || ReferenceEquals(other._tokens, null)) return false;
+            if (_tokens.Length != other._tokens.Length) return false;
+
+            for (var i = 0; i < _tokens.Length; i++)
+            {
+                if (!_tokens[i].Equals(other._tokens[i])) return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>Returns the hash code for this instance.</summary>
+        /// <returns>A 32-bit signed integer that is the hash code for this instance.</returns>
+        public override int GetHashCode()
+        {
+            var hashCode = -1248728428;
```

Comment:
add `unchecked` scope. Even if you don't need it now, someone maintaining the code may do different math